### PR TITLE
pufferpanel: build frontend from source

### PR DESCRIPTION
**Pull request by tie**
###### Description of changes

This change updates the PufferPanel package to build frontend from source. ~~Note that cross-compilation with `buildNpmPackage` seems to be broken (at least from `aarch64-linux` to `x86_64-linux`), but the frontend is platform-independent so we use `pkgsBuildBuild.buildNpmPackage`.~~

It is bundled with 2.6.7 version bump since it allows us to drop upstreamed patches.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/pufferpanel/default.nix [hunk 1]
```diff
@@ -1,43 +1,59 @@
...
-, fetchzip
 , fetchpatch
-, pathDeps ? [ ]
+, applyPatches
+, buildGoModule
+, pkgsBuildBuild # buildNpmPackage
+, makeWrapper
+, go-swag
 , nixosTests
+, pathDeps ? [ ]
```
Comment (SuperSandro2000) [inline pkgs/servers/pufferpanel/default.nix:10]: Why do we need this option? Rebuilding a package to change the PATH env in the wrapper is rather wasteful. People can just add the package to the environment.

Comment (tie) [reply pkgs/servers/pufferpanel/default.nix:10]: This was added in #128243 when package was introduced, not sure why though. If this won’t be considered a breaking change for backport, I’m all in for removing it.

Comment (SuperSandro2000) [reply pkgs/servers/pufferpanel/default.nix:10]: Should be alright in my opinion. Changing how the frontent is build as a small chance for breakage anyway.

File: pkgs/servers/pufferpanel/default.nix [hunk 2]
```diff
@@ -46,50 +61,52 @@ buildGoModule rec {
...
 
-  src = fetchFromGitHub {
-    owner = "pufferpanel";
-    repo = pname;
-    rev = "v${version}";
-    hash = "sha256-0Vyi47Rkpe3oODHfsl/7tCerENpiEa3EWBHhfTO/uu4=";
-  };
+  # Web UI is platform-independent but NodeJS fails to cross-compile at least
+  # on `aarch64-linux`-to-`x86_64-linux` builds.
+  frontend = pkgsBuildBuild.buildNpmPackage {
```
Comment (tie) [inline pkgs/servers/pufferpanel/default.nix:66]: I’m not really sure if this should be `buildPackages` or `pkgsBuildBuild`, and whether using one over another would make a difference in this case.

```
$ rg buildPackages | wc -l
1210
$ rg pkgsBuildBuild | wc -l
115
```

Apparently `pkgsBuildBuild` is not used that much?

@SuperSandro2000 could you help me with that?

File: pkgs/servers/pufferpanel/default.nix (hunk 1 above)
Comment (tie) [reply pkgs/servers/pufferpanel/default.nix]: Done.

File: pkgs/servers/pufferpanel/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/servers/pufferpanel/default.nix:66]: Those variables are for getting build or runtime versions of packages where splicing does not work aka in strings. We can take buildNpmPackage from inputs.

Comment (SuperSandro2000) [reply pkgs/servers/pufferpanel/default.nix:66]:
```suggestion
  frontend = buildNpmPackage {
```

File: pkgs/servers/pufferpanel/default.nix [hunk 3]
```diff
@@ -1,43 +1,58 @@
 { lib
-, buildGoModule
 , fetchFromGitHub
-, makeWrapper
-, fetchzip
 , fetchpatch
-, pathDeps ? [ ]
+, applyPatches
+, buildGoModule
+, pkgsBuildBuild # buildNpmPackage
```
Comment (SuperSandro2000) [inline pkgs/servers/pufferpanel/default.nix:6]:
```suggestion
, buildNpmPackage
```

**Review by SuperSandro2000 — COMMENTED** [review]
if that is not working than we have a special case herre which probably requires some more work.

**Comment by tie** [comment]
>if that is not working than we have a special case herre which probably requires some more work.

I think I’ve tracked down the cross-compilation issue to Nodejs build system. Currently `pkgsCross.gnu64.nodejs` fails because `-m64` flag is passed for both host and target toolchain on aarch64-linux.
https://github.com/nodejs/node/blob/57679e1c97a90057d5ca9e1f0d1f3c6255d20200/common.gypi#L416-L435

I can successfully build pkgsCross.gnu64.pufferpanel with the https://github.com/NixOS/nixpkgs/commit/6e59cfa66d4c50e74f1ecd2ba7146351d1f7848e commit applied.

File: pkgs/servers/pufferpanel/default.nix (hunk 3 above)
Comment (tie) [reply pkgs/servers/pufferpanel/default.nix]: Done.

**Pull request merged**